부산대 FE 김대영 - 3단계 : 단체 주문 기능 구현하기

eodudrepublic opened the pull request:
- [x] 받는 사람 기능 고도화
    - [x] 받는 사람을 최대 10명까지 등록 가능하도록 구현
    - [x] 10명의 정보가 정확하게 입력되지 않으면 정보가 반영되지 않도록 제한
    - [x] 받는 사람의 전화번호가 중복되지 않도록 구현
    - [x] 최소 선택 수량은 1개 이상 조건으로 구현
    - [x] 전화번호는 01012341234 형태로만 받도록 구현
- [x] 주문하기 페이지 고도화
- [x] 재사용 가능성이 높은 컴포넌트, Hooks가 보이면 재사용성을 고려
- [x] 주문하기 페이지의 폼을 React Hook Form를 사용해서 리팩터링 (form 데이터를 useState로 다루지 않도록 구현)

eodudrepublic committed:
refactor(app): 멘토님 피드백 기반 코드 품질 개선 및 버그 수정

eodudrepublic committed:
docs(readme): 구현할 기능 목록을 README.md에 정리

eodudrepublic committed:
feat(order): 받는 사람 기능 고도화 및 모달 UI 개선
- '받는 사람' 정보 입력을 인라인 방식에서 모달 UI로 변경

eodudrepublic committed:
feat(order): 주문 페이지 기능 고도화 및 리팩토링

Leejin-Yang commented on src/hooks/useOrderForm.ts:
이 파일은 사용하지 않는 것으로 보이는데 삭제 부탁드려요
빌드 시 오류가 발생하네요.!

Leejin-Yang commented on src/contexts/ModalContext.tsx:
모달 컨텍스트도 잘 분리해주셨네요 👍

Leejin-Yang commented on src/models/OrderFormModel.ts:
```diff
@@ -1,51 +1,44 @@
-      errors.quantity = '수량은 1개 이상이어야 합니다.';
+export const ReceiversModel = z
+  .object({
+    receivers: z.array(ReceiverModel).max(10, '최대 10명까지 추가할 수 있어요.'),
```
받는 사람 최소 기준도 필요해보여요
현재 받는 사람을 추가하지 않아도 주문이 성공하네요

Leejin-Yang commented on src/components/OrderForm.tsx:
```diff
@@ -1,107 +1,112 @@
-  totalPrice,
-}: OrderFormProps) {
+export const OrderForm = forwardRef<HTMLFormElement, OrderFormProps>(({ selectedCard }, ref) => {
+  const { receivers, setReceivers } = useOrderStore();
```
스토어에 따로 받는사람을 저장해주시는 이유가 있을까요? useForm에서 상태를 관리하고 있어서 불필요하다고 생각이 드네요

Leejin-Yang commented on src/pages/OrderPage.tsx:
```diff
@@ -20,45 +22,61 @@ export default function OrderPage() {
+
+      <VerticalSpacing size="60px" />
+
+      {createPortal(
+        <OrderButton onClick={handleOrderSubmit}>
+          {totalPrice.toLocaleString()}원 주문하기
+        </OrderButton>,
+        document.body as HTMLElement
       )}
```
주문 버튼은 포탈로 빼주신 이유가 있을까요? 폼과 관련된 컨텍스트이니 같은 DOM 트리 안에 있는게 더 어울려 보여요

Leejin-Yang commented on src/components/OrderForm.tsx:
```diff
@@ -1,107 +1,112 @@
+    methods.reset({ receivers: { receivers } });
+  }, [receivers, methods]);
+
+  useEffect(() => {
+    if (selectedCard) {
+      methods.setValue('message', selectedCard.defaultTextMessage);
+    }
+  }, [selectedCard, methods]);
```
이 부분은 개인적인 의견인데요, useEffect는 사이드 이팩트를 관리하는 훅이라 생각해 이런 상태 업데이트 관련해서는 이벤트 핸들러에 넣는 것을 추천드려요. 이 로직은 카드 클릭 이벤트에 들어갈 수 있어보여요.

useEffect가 언제 실행되는지, 그 안에서 상태가 변경되면 또 어떤 일이 일어나는지 고민해보시면 좋아보여요 😊

Leejin-Yang commented on src/hooks/useLoginForm.ts:
```diff
@@ -1,13 +1,16 @@
 export function useLoginForm() {
-  const emailInput = useInput({ validator: validateEmail });
-  const passwordInput = useInput({ validator: validatePassword });
+  const emailInput = useInput({
+    validator: validateEmail,
+  });
```
지금 로그인에서 문제가 있는데요,
입력이 없을 때 change 이벤트가 발동하지 않아 useInput의 handleChange 내 분기처리가 일어나지 않고 있어요. 그래서 기본값으로 isFormValid가 true가 되네요. 이 부분도 수정이 필요해보여요

Leejin-Yang commented on src/models/LoginFormModel.ts:
이 파일은 작성하고 사용하지 않고 있네요. 확인 부탁드려요 😊

Leejin-Yang commented on src/components/ReceiverModal.tsx:
```diff
@@ -0,0 +1,134 @@
+  const { close } = useModal();
+  const { receivers, setReceivers } = useOrderStore();
+
+  const methods = useForm<ReceiversModelType>({
+    resolver: zodResolver(ReceiversModel),
+    defaultValues: {
+      receivers,
+    },
+    mode: 'onSubmit',
+    reValidateMode: 'onSubmit',
+  });
+
+  const { fields, append, remove } = useFieldArray({
+    control: methods.control,
+    name: 'receivers',
+  });
```
주문 폼이나 이 받는 사람 폼 로직도 훅으로 분리해보면 어떨까요?

Leejin-Yang reviewed (changes requested):
대영님 안녕하세요 👋 마지막 단계까지 잘 오셨네요
몇가지 기대한 동작이 잘 작동하지 않아서 이 부분 수정이 필요해보여요
그리고 몇가지 개선할 수 있는 부분에 코멘트 달아뒀는데 확인해주시고 다시 요청 보내주세요 😊

eodudrepublic committed:
feat: 주문/로그인 기능 개선 및 코드 정리
- 주문: 받는 사람 최소 기준 유효성 검사 추가, useReceiverForm 훅 분리, createPortal 제거 후 CSS 고정 버튼 구현
- 로그인: useInput 초기 유효성 검사 반영, useLoginForm 내 로그인 로직 이동
- 코드 정리: 불필요한 파일 제거, import type 사용을 위한 ESLint 규칙 및 lint:fix 스크립트 추가

eodudrepublic commented:
> 대영님 안녕하세요 👋 마지막 단계까지 잘 오셨네요 몇가지 기대한 동작이 잘 작동하지 않아서 이 부분 수정이 필요해보여요 그리고 몇가지 개선할 수 있는 부분에 코멘트 달아뒀는데 확인해주시고 다시 요청 보내주세요 😊

수정 완료했습니다! 혹시 더 수정이 필요한 부분이 있을까요?

Leejin-Yang commented on eslint.config.js:
```diff
@@ -24,6 +24,13 @@ export default tseslint.config(
         'warn',
         { allowConstantExport: true },
       ],
+      '@typescript-eslint/consistent-type-imports': [
+        'error',
+        {
+          prefer: 'type-imports',
+          fixStyle: 'inline-type-imports',
+        },
+      ],
```
👍

Leejin-Yang commented on src/models/OrderFormModel.ts:
```diff
@@ -14,7 +14,10 @@ export const ReceiverModel = z.object({
-    receivers: z.array(ReceiverModel).max(10, '최대 10명까지 추가할 수 있어요.'),
+    receivers: z
+      .array(ReceiverModel)
+      .min(1, '받는 사람을 추가해주세요.')
```
👍

Leejin-Yang reviewed (approved):
마지막 단계까지 고생하셨습니다 👍
요구사항 잘 만족하셔서 여기서 마무리해도 될 것으로 보여요
이번 미션 리뷰는 여기까지지만 앞으로도 궁금한 점이 있다면 언제든 DM 보내주세요
고생 많았어요 😊
